x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java

```diff
@@ -159,7 +166,18 @@ protected InboundPipeline getInboundPipeline(boolean isRemoteClusterServerChanne
                 : new InboundDecoder(recycler),
             new InboundAggregator(getInflightBreaker(), getRequestHandlers()::getHandler, ignoreDeserializationErrors()),
             this::inboundMessage
-        );
+        ) {
+            @Override
+            protected void headerReceived(Header header) {
+                // TODO maybe stash the thread context
+                // start authn asynchronously
+                crossClusterAccessAuthenticationService.tryAuthenticate(header.getRequestHeaders(), ActionListener.wrap(aVoid -> {
+                    // NOOP
+                }, e -> uncaughtExceptionReference.set(new HeaderValidationException(header, e))));
```

Comment:
A few things:
* We need wrap this inside a `isRemoteClusterServerChannel` check. Right now the header check is applied to normal transport connections and cause CI failures. Alternatively, we can check `isRemoteClusterServerChannel` in the beginning of the `getInboundPipeline` method and directly return `super.getInboundPipeline` if the check is false. This would be my perference to more explicitly separate the logic, i.e. other channels do not even see any of the customisation here.
* We need let tcp handshake (`header.isHandshake()`) pass. The tcp handshake is handled completely differently and does not carry authn information. We can look into it later to see whether authn is necessary. But for now this has to be an exception.
*  We need to fail if the incoming message is a response (`header.isRequest() == false`). Response does not have authn information. A server channel should *not* handle response. So we should just fail.

Reply:
I've pushed https://github.com/elastic/elasticsearch/pull/98814/commits/1e175198f0fdc5f9a8f5bb1ad4e55ff291dc150a to guard auth:
`if (isRemoteClusterServerChannel && header.isRequest() && header.isHandshake() == false) {`

> We need to fail if the incoming message is a response (header.isRequest() == false).

I don't think this is the place to reject responses on this channel. But it can easily be done in the `InboundDecoder`.
Alternatively, we could allow authenticated **responses** go through here, and let downstream code handle responses on remote cluster server channels.
@ywangd Which one you prefer?

Reply:
Sorry, I implemented rejecting responses in the decoder: https://github.com/elastic/elasticsearch/pull/98814/commits/a48964948a98b3644a2f2d546c2d1ef8d069b220
I think keeping this override method (`#headerReceived`) only for authn is easier to reason about.

There have been a couple of minor nit changes since initially raising this.
Let me know if the final form looks good to you.

Reply:
No worries. It looks good. My main point is that responses must be rejected. I am not too concerned about exactly where they get rejected. I thought it was easier in InboundPipeline. But you found a straightforward way to incoorporate it into InboundDecoder. That's great!